std/fmt.zig

```diff
@@ -866,17 +867,39 @@ test "fmt.parseFloat" {
     _ = @import("fmt/parse_float.zig");
 }
 
-pub fn charToDigit(c: u8, radix: u8) (error{InvalidCharacter}!u8) {
-    const value = switch (c) {
-        '0'...'9' => c - '0',
-        'A'...'Z' => c - 'A' + 10,
-        'a'...'z' => c - 'a' + 10,
-        else => return error.InvalidCharacter,
-    };
+// TODO This is not inside charToDigit() due to a bug https://github.com/ziglang/zig/issues/2128#issuecomment-477877639
+const NOT = 0xff;
+const swtch = []u8{
+//  All XDigit code points in this table are in their place in this ASCII+128 table.
+//    0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,  14,  15
+    NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT,
+    NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT,
+    NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT,
+      0,   1,   2,   3,   4,   5,   6,   7,   8,   9, NOT, NOT, NOT, NOT, NOT, NOT,
+
+    NOT, 0xA, 0xB, 0xC, 0xD, 0xE, 0xF,  16,  17,  18,  19,  20,  21,  22,  23,  24,
+     25,  26,  27,  28,  29,  30,  31,  32,  33,  34,  35, NOT, NOT, NOT, NOT, NOT,
+    NOT, 0xa, 0xb, 0xc, 0xd, 0xe, 0xf,  16,  17,  18,  19,  20,  21,  22,  23,  24,
+     25,  26,  27,  28,  29,  30,  31,  32,  33,  34,  35, NOT, NOT, NOT, NOT, NOT,
+
+    NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT,
+    NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT,
+    NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT,
+    NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT,
+
+    NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT,
+    NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT,
+    NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT,
+    NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT, NOT,
+};
+
+pub fn charToDigit(c: u8, radix: u8) (error{InvalidCharacter}!u6) {
+    @import("std").debug.assert(radix <= 36);
 
+    const value = swtch[c];
```

Comment:
optimizations must come with tests

Reply:
https://github.com/ziglang/zig/issues/2128#issuecomment-477877639

Which was in the commit description. I feel some of the problems come from github's interface. I am splitting things into different commits for a reason.

Also, the way you are lowering ranged switch statements inside zig means that llvm can never optimize these sort of things. (even if it doesn't well currently)